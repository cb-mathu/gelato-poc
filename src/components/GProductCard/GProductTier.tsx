import { defineComponent, PropType } from 'vue';
import { Multiply } from '../icons';
import { Product, ProductTier } from './types';

export default defineComponent({
  props: {
    tier: {
      type: Object as PropType<ProductTier>,
    },
    product: {
      type: Object as PropType<Product>,
      required: true,
    },
    showTotal: Boolean,
  },
  setup(props) {
    const { tier, product, showTotal } = props;
    return () =>
      tier ? (
        <div class="g-product-card__tier">
          <span>
            {tier.unitPrice} <Multiply width={12} /> {tier.units} {product.labels.units}
          </span>
          &nbsp;&nbsp;&nbsp; {tier.subTotal && showTotal && <span class="g-product-card__tier-total">{tier.subTotal}</span>}
        </div>
      ) : null;
  },
});
