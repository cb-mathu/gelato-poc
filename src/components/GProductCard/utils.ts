import { ItemType, ItemTypeLabel } from './types';

export function getItemTypeLabel(entityType: ItemType | undefined): ItemTypeLabel {
  switch (entityType) {
    case ItemType.Plan:
      return ItemTypeLabel.Plan;
    case ItemType.Addon:
      return ItemTypeLabel.Addon;
    case ItemType.Adhoc:
    case ItemType.Charge:
    default:
      return ItemTypeLabel.Charge;
  }
}
