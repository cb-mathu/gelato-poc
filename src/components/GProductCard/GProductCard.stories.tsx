import { computed, ComputedRef, ref } from 'vue';
import GProductCard from './GProductCard';
import {
  ItemType,
  LoaderState,
  PricingModel,
  Product,
  ProductCardAction,
  ProductCardActionEvent,
  ProductCardActionType,
  ProductCardState,
  ProductTier,
} from './types';
import { GButton } from '../GButton';
import { getItemTypeLabel } from './utils';

export default {
  title: 'Component/GProductCard',
  component: GProductCard,
};

// Mock & Utils

const subTotal = computed(() => '$5.00');
const getQuantity = (value = 1) => computed(() => value);
const defaultDescription = 'This is enterprise level plan comes with a 14-day trial period and has a setup cost.';

const invalidProductState: ComputedRef<ProductCardState> = computed(() => {
  return {
    quantityConfig: {
      invalid: true,
      errorMessage: 'Quantity should be between 1 and 10',
      editable: true,
    },
    showTag: true,
  };
});

const actionHandler = (event: ProductCardActionEvent) => {
  console.log(`${event.action} on ${event.id}`);
};

const productAddAction: ProductCardAction = {
  id: 'element-id',
  name: ProductCardActionType.AddToCart,
  className: 'add-to-cart-btn',
  displayText: 'Add',
  element: GButton,
};

const productRemoveAction: ProductCardAction = {
  name: ProductCardActionType.Remove,
  className: 'remove-btn',
  displayText: 'Remove',
  element: 'a',
};

function productCardFactory(itemType: ItemType) {
  const pricingModel = PricingModel.FlatFee;
  switch (itemType) {
    case ItemType.Addon:
      return (id: string = 'ssl', name: string = 'SSL') =>
        new ProductCardTestBuilder(id, name, ItemType.Addon).setPrice(pricingModel, subTotal);
    case ItemType.Charge:
      return (id: string = 'day-pass', name: string = 'Day Pass') =>
        new ProductCardTestBuilder(id, name, ItemType.Charge).setPrice(pricingModel, subTotal);
    case ItemType.Plan:
    default:
      return (id: string = 'blossom', name: string = 'Blossom') =>
        new ProductCardTestBuilder(id, name, ItemType.Plan).setPrice(pricingModel, subTotal);
  }
}

const PlanCard = productCardFactory(ItemType.Plan);
const AddonCard = productCardFactory(ItemType.Addon);
const ChargeCard = productCardFactory(ItemType.Charge);
class ProductCardTestBuilder {
  product: Product;
  state: ComputedRef<ProductCardState> = computed(() => {
    return { showTag: true };
  });
  actions: ProductCardAction[] = [];
  loaderState?: LoaderState;
  constructor(id: string, name: string, type: ItemType) {
    this.product = {
      id,
      name,
      type,
      labels: {
        units: 'units',
        itemType: getItemTypeLabel(type),
      },
    };
  }

  setDescription(description?: string) {
    this.product.description = description || defaultDescription;
    return this;
  }

  setMetered() {
    this.product.metered = true;
    return this;
  }

  setPrice(pricingModel: PricingModel, subTotal: ComputedRef<string>, value?: string | ProductTier[]) {
    this.product.pricingModel = pricingModel;
    if ([PricingModel.Stairstep, PricingModel.Tiered, PricingModel.Volume].includes(pricingModel)) {
      this.product.tiers = computed(() => value as ProductTier[]);
    }
    this.product.subTotal = subTotal;
    PricingModel.PerUnit === pricingModel && (this.product.unitPrice = computed(() => value as string));
    this.product.quantity = PricingModel.FlatFee !== pricingModel ? getQuantity() : undefined;
    return this;
  }

  setActions(actions: ProductCardAction[]) {
    this.actions = actions;
    return this;
  }

  setRemoveAction() {
    this.actions.push(productRemoveAction);
    return this;
  }

  setQuantity(quantity: ComputedRef<number> = getQuantity()) {
    this.product.quantity = quantity;
    return this;
  }

  setEditable() {
    this.state.value.quantityConfig = {
      ...(this.state.value.quantityConfig || {}),
      editable: true,
    };
    return this;
  }

  setLoading() {
    this.state.value.quantityConfig = {
      ...(this.state.value.quantityConfig || {}),
      editable: true,
    };
    this.loaderState = LoaderState.loading;
    return this;
  }

  setReadonly() {
    this.state.value.quantityConfig = {
      ...(this.state.value.quantityConfig || {}),
      editable: false,
    };
    this.loaderState = LoaderState.readonly;
    return this;
  }

  setInvalid() {
    this.state.value.quantityConfig = {
      ...(this.state.value.quantityConfig || {}),
      ...invalidProductState.value.quantityConfig,
    };
    return this;
  }

  data() {
    return {
      product: this.product,
      actions: this.actions,
    };
  }

  getComponent() {
    const key = Math.random();
    return () => (
      <GProductCard
        key={key}
        onAction={actionHandler}
        loaderState={this.loaderState}
        product={this.product}
        actions={this.actions}
        state={this.state}
        labels={{ quantity: 'units', seeDetails: 'See details', hideDetails: 'Hide details' }}
      />
    );
  }
}

const Template = () => ({
  setup() {
    const sampleProduct: Product = {
      id: 'cbdemo_scale',
      name: 'Scale',
      description: defaultDescription,
      quantity: computed(() => 2),
      subTotal: computed(() => '$10.00'),
      unitPrice: computed(() => '$5.00'),
      type: ItemType.Plan,
      labels: {
        units: 'units',
        itemType: getItemTypeLabel(ItemType.Plan),
      },
    };

    return () => (
      <div style="display: grid; grid-gap: 1rem;">
        <GProductCard
          onAction={actionHandler}
          product={sampleProduct}
          actions={[productAddAction]}
          state={computed(() => {
            return { quantityConfig: { editable: true }, showTag: true };
          })}
          labels={{ quantity: 'units', seeDetails: 'See details', hideDetails: 'Hide details' }}
        />
        <GProductCard
          onAction={actionHandler}
          product={sampleProduct}
          actions={[productRemoveAction]}
          state={invalidProductState}
          labels={{ quantity: 'units', seeDetails: 'See details', hideDetails: 'Hide details' }}
        />
      </div>
    );
  },
});

export const Story = Template.bind({});

export const DefaultPlanCard = PlanCard().getComponent();
export const DefaultAddonCard = AddonCard().getComponent();
export const DefaultChargeCard = ChargeCard().getComponent();
export const CardWithDescription = PlanCard().setDescription().getComponent();
export const CardWithAction = PlanCard().setRemoveAction().getComponent();
export const MeteredCard = PlanCard().setMetered().getComponent();
export const PerUnitCard = PlanCard().setPrice(PricingModel.PerUnit, subTotal, '$5').getComponent();
export const EditableUnitPricingCard = PlanCard().setPrice(PricingModel.PerUnit, subTotal, '$5').setEditable().getComponent();
export const InvalidUnitPricingCard = PlanCard().setPrice(PricingModel.PerUnit, subTotal, '$5').setInvalid().getComponent();
export const TierPricingCard = PlanCard()
  .setPrice(PricingModel.Tiered, subTotal, [
    {
      units: 2,
      unitPrice: '$1.00',
      subTotal: '$2.00',
    },
    {
      units: 3,
      unitPrice: '$1.00',
      subTotal: '$3.00',
    },
  ])
  .setQuantity(getQuantity(5))
  .getComponent();

export const EditableTierPricingCard = PlanCard()
  .setPrice(PricingModel.Tiered, subTotal, [
    {
      units: 2,
      unitPrice: '$1.00',
      subTotal: '$2.00',
    },
    {
      units: 3,
      unitPrice: '$1.00',
      subTotal: '$3.00',
    },
  ])
  .setDescription()
  .setQuantity(getQuantity(5))
  .setEditable()
  .getComponent();

export const InvalidTierPricingCard = PlanCard()
  .setPrice(PricingModel.Tiered, subTotal, [
    {
      units: 2,
      unitPrice: '$1.00',
      subTotal: '$2.00',
    },
    {
      units: 3,
      unitPrice: '$1.00',
      subTotal: '$3.00',
    },
  ])
  .setQuantity(getQuantity(5))
  .setDescription()
  .setEditable()
  .setInvalid()
  .getComponent();

export const LoadingTierPricingCard = PlanCard()
  .setPrice(PricingModel.Tiered, subTotal, [
    {
      units: 2,
      unitPrice: '$1.00',
      subTotal: '$2.00',
    },
    {
      units: 3,
      unitPrice: '$1.00',
      subTotal: '$3.00',
    },
  ])
  .setQuantity(getQuantity(5))
  .setDescription()
  .setEditable()
  .setLoading()
  .getComponent();

export const ReadonlyTierPricingCard = PlanCard()
  .setPrice(PricingModel.Tiered, subTotal, [
    {
      units: 2,
      unitPrice: '$1.00',
      subTotal: '$2.00',
    },
    {
      units: 3,
      unitPrice: '$1.00',
      subTotal: '$3.00',
    },
  ])
  .setQuantity(getQuantity(5))
  .setDescription()
  .setReadonly()
  .setRemoveAction()
  .getComponent();
