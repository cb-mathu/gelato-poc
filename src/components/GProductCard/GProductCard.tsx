import { computed, ComputedRef, defineComponent, PropType } from 'vue';
import GTag from '../GTag/GTag';
import GProductCardQuantity from './GProductCardQuantity';
import TierRow from './GProductTier';
import {
  ItemType,
  LoaderState,
  Product,
  ProductCardAction,
  ProductCardActionEvent,
  ProductCardActionType,
  ProductCardLabels,
  ProductCardState,
} from './types';
import { TagVariant } from '../GTag';
import { isNotEmpty } from '../../utils/common';
import { GInputErrorMessage } from '../GInput';
import { Clickable } from '../../utils/accessibility';
import { useToggle } from '../../composables';
import { GSlideTransition } from '../animations/GSlideTransition';

export default defineComponent({
  name: 'ProductCard',
  props: {
    state: {
      type: Object as PropType<ComputedRef<ProductCardState>>,
      default: () => ({
        showTag: true,
      }),
    },
    product: {
      type: Object as PropType<Product>,
      required: true,
    },
    actions: {
      type: Array as PropType<Array<ProductCardAction>>,
      required: false,
    },
    onAction: {
      type: Function as PropType<(event: ProductCardActionEvent) => void>,
      required: false,
    },
    loaderState: {
      type: String as PropType<LoaderState>,
    },
    labels: {
      type: Object as PropType<ProductCardLabels>,
      required: true,
    },
    recommended: {
      type: Boolean,
      required: false,
    },
  },
  emits: ['action'],
  setup(props, { emit }) {
    return () => {
      const { product, state, actions, labels } = props;

      const quantityConfig = state?.value?.quantityConfig;

      const loaderState = props?.loaderState;

      const isLoading = computed(() => props.loaderState === LoaderState.loading);

      const [showTiers, toggleTiers] = useToggle(false);

      const isPlan = computed(() => product.type === ItemType.Plan).value;

      const Tag = () =>
        product.labels.itemType ? (
          <GTag variant={isPlan ? TagVariant.Primary : TagVariant.Secondary}>{product.labels.itemType}</GTag>
        ) : null;

      const Heading = () => <h4 class="g-product-card__heading">{product.name}</h4>;

      const Description = () => <p class="g-product-card__desc">{product.description}</p>;

      const Actions = () => (
        <div class="g-actions">
          {actions?.map(({ element, displayText, id, name: action, className }) => {
            return (
              <Clickable
                as={element}
                id={id}
                class={className}
                onClick={() => emit('action', { action, id: product.id })}
                disabled={Boolean(loaderState)}
              >
                {displayText}
              </Clickable>
            );
          })}
        </div>
      );

      const Quantity = () => {
        const quantityUpdateHandler = (value: number) => {
          emit('action', { action: ProductCardActionType.UpdateQuantity, id: product.id, value });
        };

        const qtyListeners: any = {
          'onUpdate:modelValue': quantityUpdateHandler,
        };
        return (
          <>
            <div class="g-product-card__qty">
              {!props.state.value?.quantityConfig?.hide && (
                <GProductCardQuantity
                  product={product}
                  labels={labels}
                  state={state.value}
                  loaderState={loaderState}
                  showTiers={showTiers}
                  isLoading={isLoading}
                  toggleTiers={toggleTiers}
                  qtyListeners={qtyListeners}
                />
              )}
              {product.metered && <span>⚡️ &nbsp; Charged based on the usage</span>}

              {props.state.value?.quantityConfig?.hide && props.state.value?.quantityConfig?.hasQuantity && (
                <span class="g-product-card__qty__text">
                  for {product.quantity} {product.labels.units}
                </span>
              )}
            </div>
          </>
        );
      };
      const Price = () => {
        return (
          <>
            <span class={props.state.value?.quantityConfig?.hide ? 'g-product-card__amount__left' : 'g-product-card__amount'}>
              {product?.subTotal?.value}
            </span>
          </>
        );
      };

      const TierBreakdown = () => {
        return (
          <div>
            <GSlideTransition duration={250}>
              {(product.tiers?.value || []).length > 1 && showTiers.value && (
                <div>
                  <div class="g-product-card__tiers" key="shown">
                    {product.tiers?.value.map((tier) => (
                      <TierRow tier={tier} showTotal={true} product={product} />
                    ))}
                  </div>
                </div>
              )}
            </GSlideTransition>
          </div>
        );
      };
      const classList = {
        [`g-card g-product-card`]: true,
        [`g-card-recommended`]: props.recommended,
        [`g-product-card--loading`]: isLoading.value,
      };
      const qtyPriceClassList = {
        [`g-card__qty-margin`]: !props.state.value?.quantityConfig?.hide || undefined,
        [`g-card__right`]: true,
      };
      return (
        <section class={classList} id={`g-product-card__${product.id}`}>
          <div class="g-card__content">
            {state?.value?.showTag && <Tag />}
            <div class="g-card__content_row">
              <div class="g-card__left">
                <Heading />
                {isNotEmpty(product.description) && <Description />}
                <Quantity />
              </div>
              <div class={qtyPriceClassList}>
                <Price />
              </div>
            </div>

            {quantityConfig?.invalid && quantityConfig.errorMessage && <GInputErrorMessage errorMessage={quantityConfig.errorMessage} />}
            {isNotEmpty(product.tiers) && <TierBreakdown />}
            {isNotEmpty(actions) && <Actions />}
          </div>
        </section>
      );
    };
  },
});
