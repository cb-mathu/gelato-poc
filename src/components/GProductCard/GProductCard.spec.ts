import { render, screen, fireEvent } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import * as stories from './GProductCard.stories';

const {
  DefaultAddonCard,
  DefaultChargeCard,
  DefaultPlanCard,
  InvalidTierPricingCard,
  InvalidUnitPricingCard,
  CardWithDescription,
  CardWithAction,
  MeteredCard,
  PerUnitCard,
  EditableTierPricingCard,
  EditableUnitPricingCard,
} = stories;

describe('GProductCard : ', () => {
  test('renders default plan card', () => {
    const { container } = render(DefaultPlanCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders default addon card', () => {
    const { container } = render(DefaultAddonCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders default charge card', () => {
    const { container } = render(DefaultChargeCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders metered card', () => {
    const { container } = render(MeteredCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders card with description', () => {
    render(CardWithDescription);
    expect(screen.getByText('This is enterprise level plan comes with a 14-day trial period and has a setup cost.')).toBeVisible();
  });

  test('renders per unit card', () => {
    const { container } = render(PerUnitCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders card as readonly state when state is set', () => {
    const { container } = render(PerUnitCard);
    const qty = container.querySelector('input');
    const initialValue = qty?.value;
    userEvent.type(qty as HTMLElement, '1');
    expect(qty?.value).toBe(initialValue);
  });

  test('renders per unit card in editable state', () => {
    const { container } = render(EditableUnitPricingCard);
    const qty = container.querySelector('input');
    const initialValue = qty?.value;
    userEvent.type(qty as HTMLElement, '1');
    expect(qty?.value).not.toBe(initialValue);
  });

  test('renders tiered card in editable state', () => {
    const { container } = render(EditableTierPricingCard);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders card with actions passed', () => {
    render(CardWithAction);
    expect(screen.getByText('Remove')).toBeVisible();
  });

  test('show/hide tiers for tiered card', async () => {
    const { container } = render(EditableTierPricingCard);
    await fireEvent.click(screen.getByText('See details'));
    const el: HTMLElement | null = container.querySelector('.g-product-card__tier');
    expect(el?.textContent).toEqual('$1.00  2 units    $2.00');
    await fireEvent.click(screen.getByText('Hide details'));
    expect(screen.getByText('See details')).toBeVisible();
  });

  describe('Invalid : ', () => {
    test('renders per unit card in invalid state', () => {
      const { container } = render(InvalidUnitPricingCard);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders tiered card in invalid state', () => {
      const { container } = render(InvalidTierPricingCard);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
