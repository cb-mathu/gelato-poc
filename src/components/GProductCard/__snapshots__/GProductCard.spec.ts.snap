// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GProductCard :  Invalid :  renders per unit card in invalid state 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          
          <span
            class="g-product-card__unit-price"
          >
            $5
             
            <svg
              fill="currentColor"
              viewBox="0 0 20 20"
              width="20px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          
          <div
            class="g-quantity g-quantity--invalid"
          >
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__decrement"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
            <input
              aria-label="units"
              aria-live="assertive"
              autocomplete="off"
              class="g-quantity__element"
              id="GQuantity_4"
              invalid="true"
              modelmodifiers="[object Object]"
              modelvalue="1"
              name=""
              placeholder=""
              type="number"
            />
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__increment"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <span
            class="qty_label"
          >
            units
          </span>
          
          
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <div>
      <div
        class="g-notification g-notification--alert g-notification--small"
      >
        <svg
          class="g-notification__icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.707 7.293a1 1 0 00-1.414 1.414L8.586 10l-1.293 1.293a1 1 0 101.414 1.414L10 11.414l1.293 1.293a1 1 0 001.414-1.414L11.414 10l1.293-1.293a1 1 0 00-1.414-1.414L10 8.586 8.707 7.293z"
            fill-rule="evenodd"
          />
        </svg>
        <p
          class="g-notification__message"
        >
          
          <span>
            Quantity should be between 1 and 10
          </span>
          
        </p>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  Invalid :  renders tiered card in invalid state 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <p
          class="g-product-card__desc"
        >
          This is enterprise level plan comes with a 14-day trial period and has a setup cost.
        </p>
        
        <div
          class="g-product-card__qty"
        >
          
          
          <div
            class="g-quantity g-quantity--invalid"
          >
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__decrement"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
            <input
              aria-label="units"
              aria-live="assertive"
              autocomplete="off"
              class="g-quantity__element"
              id="GQuantity_5"
              invalid="true"
              modelmodifiers="[object Object]"
              modelvalue="5"
              name=""
              placeholder=""
              type="number"
            />
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__increment"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <span
            class="qty_label"
          >
            units
          </span>
          
           Â 
          <a
            aria-disabled="false"
            as="a"
            class="g-tier-details-toggle"
            disabled="false"
            role="button"
            tabindex="0"
          >
            See details
          </a>
          
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <div>
      <div
        class="g-notification g-notification--alert g-notification--small"
      >
        <svg
          class="g-notification__icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.707 7.293a1 1 0 00-1.414 1.414L8.586 10l-1.293 1.293a1 1 0 101.414 1.414L10 11.414l1.293 1.293a1 1 0 001.414-1.414L11.414 10l1.293-1.293a1 1 0 00-1.414-1.414L10 8.586 8.707 7.293z"
            fill-rule="evenodd"
          />
        </svg>
        <p
          class="g-notification__message"
        >
          
          <span>
            Quantity should be between 1 and 10
          </span>
          
        </p>
      </div>
    </div>
    <div>
      <transition-stub>
        
        <!---->
        
      </transition-stub>
    </div>
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders default addon card 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__ssl"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--secondary g-product-card__type"
      variant="secondary"
    >
      
      Addon
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          SSL
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          <!---->
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders default charge card 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__day-pass"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--secondary g-product-card__type"
      variant="secondary"
    >
      
      One-time
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Day Pass
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          <!---->
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders default plan card 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          <!---->
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders metered card 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          <!---->
          <span>
            â¡ï¸ Â  Charged based on the usage
          </span>
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders per unit card 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <!---->
        
        <div
          class="g-product-card__qty"
        >
          
          <span
            class="g-product-card__unit-price"
          >
            $5
             
            <svg
              fill="currentColor"
              viewBox="0 0 20 20"
              width="20px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          
          <span
            class="g-product-card__qty__text"
          >
            1
          </span>
          <span
            class="qty_label"
          >
            units
          </span>
          
          
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`GProductCard :  renders tiered card in editable state 1`] = `
<section
  class="g-card g-product-card"
  id="g-product-card__blossom"
>
  <div
    class="g-card__content"
  >
    <span
      class="g-tag g-tag--primary g-product-card__type"
      variant="primary"
    >
      
      Plan
      
    </span>
    <div
      class="g-card__content_row"
    >
      <div
        class="g-card__left"
      >
        <h4
          class="g-product-card__heading"
        >
          Blossom
        </h4>
        <p
          class="g-product-card__desc"
        >
          This is enterprise level plan comes with a 14-day trial period and has a setup cost.
        </p>
        
        <div
          class="g-product-card__qty"
        >
          
          
          <div
            class="g-quantity"
          >
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__decrement"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
            <input
              aria-label="units"
              aria-live="assertive"
              autocomplete="off"
              class="g-quantity__element"
              id="GQuantity_2"
              modelmodifiers="[object Object]"
              modelvalue="5"
              name=""
              placeholder=""
              type="number"
            />
            <button
              aria-disabled="false"
              as="button"
              class="g-quantity__change g-quantity__increment"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <span
            class="qty_label"
          >
            units
          </span>
          
           Â 
          <a
            aria-disabled="false"
            as="a"
            class="g-tier-details-toggle"
            disabled="false"
            role="button"
            tabindex="0"
          >
            See details
          </a>
          
          <!---->
          <!---->
        </div>
        
      </div>
      <div
        class="g-card__qty-margin g-card__right"
      >
        
        <span
          class="g-product-card__amount"
        >
          $5.00
        </span>
        
      </div>
    </div>
    <!---->
    <div>
      <transition-stub>
        
        <!---->
        
      </transition-stub>
    </div>
    <!---->
  </div>
</section>
`;
